17 fvm mod updates for limiters venkata nishikawa and barth

rlfontenot opened the pull request:
This issue/pull request pulls in the non-MLP limiters from the CFD Research development version. In the CFD Research dev version, much of the limiter rules are functionalized to reduce repeated code and ease adding in new limiters, which in this issue is the Nishikawa higher-order family of limiters. Nishikawa's limiter leverages much of Venkatakrishnan's limiter and follows a similar form, hence the functionalization so the 'recipe'/rule looks the same for both limiters except for the function. The 'epsilon' function is also pulled out of the limiter rules to reduce repeatability and enable overriding of those functions in the future.  The Barth/Nodal Barth function was also made into a function and those rules modified for this purpose. 

Additionally, since we always access the same data (but in different functions) max/min's, those functions are reworked to store as a store<pair<real,real>>, with additional filtering functions (spatialFilter.loci) to enable the user the ability to modify how the max/min's and averages are computed (averages being needed for Barth and future MLP). Verified with chem example cases. Code in the limiter and added functions/rules were commented with Doxygen.

Commit:
RLF[#17]: initial commit moving variables to Loci header. Moving limiter functions to support_header to help aid in simplifying code to incorporate new limiters

Commit:
RLF[#17]: Functions declared for Barth, Venka, and Nishikawa in limiter_support.cc, additional comments

Commit:
RLF[#17]: Implemented the Nishikawa limiter (3-5th order, 2nd order works as well). Works in same manner as Venka limtiers (face, cell, and special stencil). Venka factor and max/min/norm (per stencil) now put into own rules to simpilfy limiter rules using new specialized variables (struct) as these are always accessed at the same time--so now they are contiguous in memory.

Commit:
RLF[#17]: compile/install fixes for FVMMod/limiters. Zero-limiter sensitivites as per CFDRC NASA Phase II reporting on instabilities caused by those values. Nodal/Barth cell max/min's similar to Venka/Nish limiter max/min structure, which includes spatialFiltering for cell2node max/mins (and others). .gitignore added as well to exclude compile-time files.

Commit:
RLF[#17]: missing nisPow default rule in limiter.loci. tested in chem's quicktest and cylinder cases in examples for all limiters

Commit:
RLF[#17]: fix formatting issue

Commit:
RLF[#17]: fix for nisPow=4

TimothyEDawson commented:
@EdwardALuke Could you take a look at this pull request to see if it needs anything?

EdwardALuke committed:
Merge branch 'dev' into 17-fvm-mod-updates-for-limiters-venkata-nishikawa-and-barth

EdwardALuke commented:
Yes, I have been looking over the code.  There is quite a bit to this and
there are a number of concerns that I have that range broadly.  Some of
them have performance implications, some of them relate to good practice
for libraries (which I also violated in the FVMOverset header files)  Some
other topics that we probably need to have a  broader discussion about such
as the documentation standards for Loci rules.    I am trying to complete
the review so that we can have a discussion about it maybe sometime in the
next few days.

-Ed


On Fri, Jan 3, 2025 at 11:59 AM Tim Dawson ***@***.***> wrote:

> @EdwardALuke <https://github.com/EdwardALuke> Could you take a look at
> this pull request to see if it needs anything?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rlfontenot/loci/pull/43#issuecomment-2569613357>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADFTIPAD46QSIBXPE5BD5UL2I3FZ7AVCNFSM6AAAAABTSDU7DWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKNRZGYYTGMZVG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

EdwardALuke committed:
Fixed issues after merging in recent dev changes

EdwardALuke commented:
Some of the performance related concerns relate to separating out the small functions to a separate limiter_support.cc file.  For example, there was a lightweight function called vlimit that was an inline function that was contained within the limiter.loci file and now and as a result can no longer be inlined.  Any small function that works on only one item at a time (as opposed to an array of items) should more than likely be an inlined function.  If you must move this to a separate file, then it needs to be moved to a header file and made inline.   Also I note that the implementation of this function has changed and I don't think that the regression tests are going to be sufficient to properly exercise this change, so with this change I think some more extensive production runs need to be run to verify that this does not impact the robustness of the code as this is the default limiter, we need to be more careful about this change.  

Another problem I see is in the implementation of the nis_limit function.  In this function there are the nisPow argument is passed in as a double but there are later lines such as:
 if (nisPow == 3)
This type of conditional test is likely not to behave as expected.  The conversion of integer to floating point representation is not guaranteed to be exact and so an equals comparison is not going to lead to alot of unexpected surprises.  If this is coded this way then nisPow should probably be typed as an integer.  Is it necessary that it is a floating point number?

EdwardALuke commented:
A reference about how some abstraction techniques can impact performance can be found at
https://www.computerenhance.com/p/clean-code-horrible-performance

rlfontenot commented:
@EdwardALuke to your point about nisPow: that should be an int; I use it that way and there's not a good reason to store it as a float/real. And yes, I'm aware of the real == int not behaving properly...that was a mistake. I will make those changes. As to inlining, no problem moving those function to the headers; I can make that change as well.

Commit:
RLF[#17]: Changed nisPow from real to int to fix real-to-int comparison. Moved limiter functions to header and changed to inline for performance

EdwardALuke commented:
Ok, so lets talk a bit about the differences between Loci and an application.  The Loci framework is a library and so it has internal and external interfaces.  So we need to be careful about some issues with Loci that we don't have to think about for an application like chem or GGFS.  Mostly this is about the header files, we need to decide what should be part of the public API and what is internal.  This is important for code documentation, so one thing we need to ask about what goes in the public header files is:  Do we want to support backward compatibility for this function because an end-users were using this API?    You will notice that I have header files in the System directory, in the lpp directory, etc.  These are there because they are part of the internal API.  I don't want to get stuck supporting backward compatibility for those API's, I want to be able to change them as needed to extend functionality.    I would argue that functions such as the vlimit should be part of the internal API and thus I would recommend that the header files be part of the FVMMod directory because they are an internal API.   Now the class for storing max and min values maybe should be part of the public interface, but at any rate, a case needs to be made for if something should be part of the public API or internal API and the header files placed appropriately.  

Another issue is namespace pollution.  For the library we don't want to import symbols into the global namespace, this is something that we leave for the user to decide.  Once you have a using std::cerr in one of these header files for example, then it is now forced to be in the global namespace even if it might conflict with an end user symbols (see recent discussions on Map for example).    The include<Loci.h> does this, but I have included this for C++ novices to include so they don't need to learn about C++ namespaces.  For more sophisticated users the #include <Loci> does everything that #include <Loci.h> does except it doesn't pollute the global namespace.  I see that you have several header files that have this problem (note I also made this error in the FVMOverset include files so that needs to be fixed as well).

Finally, I noticed that you are have #include directives in the .lh files.   I have been trying to keep the C preprocessor actions and the Loci preprocessor actions separate.  At least for me it is helpful for me to think of these as two separate issues, what do I need to include for the Loci part and what do I need to include for the C++ support.   I would like to hear your arguments for having #include directives in .lh files.  I am open, but if there isn't a compelling reason my preference would be to avoid this in library APIs.  

One other issue about moving the functions out of the .loci file could be a consideration for a future where automatic gpgpu kernels are being generated, right now you would not be able to call these functions from a gpgpu kernel without some additional cuda specific annotations.  I am thinking in the long term that we might want the loci preprocessor to define these so that the cuda coding can be completely hidden from the user (which would allow us to switch to other non-proprietary accelerator languages).  This isn't a concern right now and for limited use would probably be easy to change in the future, but it is something you should be aware of.

rlfontenot commented:
Let's start with headers; I'm inclined to agree with removing header inclusions in the .lh files. I've noticed when doing the FAD work that this can be problematic specifically to your point about namespace pollution--it just causes more headaches than it actually solves. Besides, the lpp can sort out the types without any problem, and any headers needed can be included in the .loci, .h, or .cc/.cpp files. That's a simple thing for me to remove and/or replace with #include <Loci>, which would need to be done with the rest of the .loci files in FVMMod that I didn't modify for this issue--and since that should be done for FVMOverset as well, we should open a separate issue for changing #include <Loci.h> to #include <Loci> for both of these modules.

With the max/min -- limiter functions: I almost split those into separate headers with my latest push, but kept them together for continuity. I can get behind the argument on public/private API usage--as a limiter function is generally going to be within the current module, while a max/min/norm operation is much more pervasive. That does raise the question though of if those types of operations should be within a given private (internal) module (FVM for example) or be elsewhere. I.E. beyond usage, for example using grad(X) or limiter(X) or even cellmaxmin(X), does that operation need to be opened up more broadly to users/developers?
On a side note, I noticed that during compilation $type X store<real> (and others) for the limiter, etc. parametrics threw all kinds of warnings about retyping. Since we have previously not had that kind of warning, and often repeat those declarations when writing rules for different store types, what should we do to remove those warnings?

To your last point, that's something to think about and something that would probably make sense to move that back to the .loci. That probably warrants further discussion as we move in that direction.

rlfontenot commented:
I also appreciated the 'clean' code article and performant code, since we had been moving to the clean paradigm without necessarily thinking of performance ramifications.

EdwardALuke commented:
Part of the design of Loci was to provide a way to abstract code (clean code goals) without the performance hit.  Alot of thought about these issues have gone into it behind the scenes.

EdwardALuke commented:
With regards to the new warnings about retyping variables, this was implemented with the recent changes to consider the Loci dependencies.  One thing that I changed in that is automatic guarding of $include files.  In the new lpp if a $include file is encountered a second time then it doesn't load it a second time.  For the most part this shouldn't cause a problem, but it may change behavior if a type was redefined, then the order of includes matter.  So I added a warning. The warning is only issued if there is a second $type and the type specified was different.  To remove the warnings you can add a $untype variable command before the $type variable which tells lpp that you intended to retype the variable.   In general we probably need to be better about moving the type declaratons to .lh files so that we are less likely to type something in a conflicting way.  This also is important in the new documentation features since documentation associated with the type gets thrown away on retyping and so it makes automatic document generation problematic.  With this regard we may want to switch to a different paradigm that avoids retyping.  For example we might want to use grads(S), then gradv3d(V), and gradv(M) to avoid retyping.  This will also help with the documentation.  Also in this case the type of the S, V, and M should probably be in the .loci file as we don't want this to conflict with a user that just wants to create a variable called V (similar to the name pollution problem I mentioned before).

rlfontenot commented:
Using S,V,M would for scalar, vect3d, and vector parametrics within the .loci files would be a nice standard to implement now rather than having to retype the variable and necessitating untype.

EdwardALuke commented:
Also regarding the generalization of the max,min facilities that may be of more general interest.  I think it still belongs in the FVMMod as the concept of nodes and cells is tied to the FVM mesh.  But it could be moved and documented outside of the limiter functionality.  There are some other things in misc.loci that could be argued are even more general than FVM, such as the fileNumber(X) or the L1Norm(X).  But the reality is that we can always refine the API, probably these are starting to get into the lower priority aspects.

EdwardALuke commented:
Ok, another problem I see is the MaxMinv operator definition that actually has empty code.  This is very bad.  When Loci schedules the problem on multiple processors it is going to use this operator to combine the partial results from a neighboring processor.  You cannot just not supply this because the rule internally didn't use join, Loci is going to use this to do parallel joins.  So the way this is coded it will only work properly on a single processor.

rlfontenot commented:
Thanks for pointing that out. When I implemented that, I could not get the join operator to work property. Now with a little more experience, I will rework that.

EdwardALuke commented:
Ok, you can look at the Maximum template in rule.h for guidance, e.g.

 template <class T> struct Maximum<Vect<T> > {
    template <class U> void operator()(Vect<T> &res ,const U &arg)
    {
      int vs = res.getSize() ;
      for(int i=0;i<vs;++i)
        res[i] = max(res[i],arg[i]) ;
    }
  } ;

Commit:
RLF[#17]: corrected join operator for maxmin struct, removed Loci.h from headers/preambles.

wandadars commented:
[only semi-on-topic] One issue that I had when I wanted to implement the MLP and MLP-pw limiter in the Loci was that the MLP-pw limiter for example uses a pressure sensor, so there would need to be some sort of way to allowing a code that is using the Loci-implemented limiter to provide things like cell values of the scalar. I just wasn't well versed enough in Loci to figure it out and ended up having to wholly in stream. Is that something that isn't that challenging from your perspectives?

EdwardALuke commented:
As for the MLP-pw style of limiter, there are several ways that such a functionality could be provided by the FVMMod, but I think implementing that in the solver is probably the best option.  Really it seems like this would involve creating a new argument to the parametric variable that would capture the weighting value.   But this would require solvers to change their implementation to pass in the extra parameter.  Since there doesn't seem to be a general foundational principle to this, the best approach to supporting this functionality in a user function is probably just have a module that you load in to override the default limiter that is designed for the solver.  The FVMMod is more for general purpose tools that are agnostic to what the code is doing (e.g. maybe it is used in a code that solves equations that don't even have the straightforward notion of pressure, like the radiation or boltzmann solvers).

rlfontenot commented:
Yeah, I agree, since the 'better' MLP limiters are some function of the additional pressure term. It ends up requiring some modification in both Loci and the general solvers, which is not what we want. That's why I ended up removing the actual implementation of the MLP limiter from this issue. The underlying structure is still there, so you can call the proper averages (which you can justify being useful in other contexts), but the limiter itself shouldn't go here.

wandadars commented:
Thanks for insight @EdwardALuke  and @rlfontenot .  That makes sense to draw the line between the machinery that a specific limiter might need and the code-specific values that would utilize that machinery. While the MLP was certainly fun to implement with how it pulls in various stencils of data, it would definitely be nice to offload those generic implementation details onto Loci.

Commit:
RLF[#17]: merge update from dev

EdwardALuke committed:
Merge remote-tracking branch 'origin/dev' into 17-fvm-mod-updates-for-limiters-venkata-nishikawa-and-barth

rlfontenot commented:
Appreciate the merge from dev, that was next on my to-do list.

EdwardALuke committed:
fixed mangled commit on file

EdwardALuke commented:
I also just fixed a borked file probably due to an unintentional editor mishap.  Are you available to discuss these changes.  There are a number of issues that probably needs to have a more interactive discussion than comments can support.

rlfontenot commented:
Yep, I have time now. Let me send you a teams link.
Raymond




________________________________
From: Edward A Luke ***@***.***>
Sent: Friday, March 7, 2025 10:27 AM
To: EdwardALuke/loci ***@***.***>
Cc: Raymond Fontenot ***@***.***>; Mention ***@***.***>
Subject: [EXTERNAL] Re: [EdwardALuke/loci] 17 fvm mod updates for limiters venkata nishikawa and barth (PR #43)


I also just fixed a borked file probably due to an unintentional editor mishap. Are you available to discuss these changes. There are a number of issues that probably needs to have a more interactive discussion than comments can support.

—
Reply to this email directly, view it on GitHub<https://github.com/EdwardALuke/loci/pull/43#issuecomment-2706888522>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/A4XOISVMJCM7D2E3CNINU5L2THCILAVCNFSM6AAAAABTSDU7DWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMBWHA4DQNJSGI>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

[EdwardALuke]EdwardALuke left a comment (EdwardALuke/loci#43)<https://github.com/EdwardALuke/loci/pull/43#issuecomment-2706888522>

I also just fixed a borked file probably due to an unintentional editor mishap. Are you available to discuss these changes. There are a number of issues that probably needs to have a more interactive discussion than comments can support.

—
Reply to this email directly, view it on GitHub<https://github.com/EdwardALuke/loci/pull/43#issuecomment-2706888522>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/A4XOISVMJCM7D2E3CNINU5L2THCILAVCNFSM6AAAAABTSDU7DWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMBWHA4DQNJSGI>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

rlfontenot commented:
@EdwardALuke working on this now; if you look at the development version of limiter.loci, should the vlimit function still take the ref values (line 844, for example), even though the function ifdef (line 773) to remove it from the vlimit function was put there 13 years ago? 

Also, and maybe this is one of those more philosophical points, is it really more performant to compute for every cell per every subiteration the quantity Kl3 (line 836) rather than store that, since that's effectively just a param that is constant everywhere?

Since we've got the realToDouble for all of the limiters, and as it looks like from both of our tests that sensitivities pushed through the limiters doesn't make any sense, should we (and this may be better served in a separate issue), look to just making the limiter calculations purely double, vect3d<double>, and vector<double>, so when in the FAD mode we don't waste extra computations and memory carrying around sensitivities that we just discard anyways?

EdwardALuke commented:
The #ifdef could be removed and we could potentially remove ref as an argument to vlimit.  Since the code is inlined this should not have any performance implication, but if ref is not referenced in vlimit at all (after removing the #ifdef lines, then it could be confusing as to why it was passed in.

As for the Kl3 line the optimizing compilers should be able to hoist this out of the loop.  Loci provides the aliasing hints to the allow the compiler to do this, but it might be worth checking to make sure it actually does.

As for storing the limiters as double, etc.   Yes I think that is probably what we want to do once we have made the commitment that this is what we always want to do.  I think we are probably fairly certain at this point.

Commit:
RLF[#17]: refactor of updates back to limiter standards. split limiters into separate files within FVMMod/limiter directory. will need to run unit tests next
